common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -396,6 +403,56 @@ public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
     }
   }
 
+  @Override
+  public void removeShuffleMerge(RemoveShuffleMerge msg) {
+    AppShuffleInfo appShuffleInfo = validateAndGetAppShuffleInfo(msg.appId);
+    if (appShuffleInfo.attemptId != msg.appAttemptId) {
+      throw new IllegalArgumentException(
+          String.format("The attempt id %s in this RemoveShuffleMerge message does not match "
+                  + "with the current attempt id %s stored in shuffle service for application %s",
+              msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
+    }
+    appShuffleInfo.shuffles.computeIfPresent(msg.shuffleId, (shuffleId, mergePartitionsInfo) -> {
+      boolean deleteCurrent =
+          msg.shuffleMergeId == DELETE_CURRENT_MERGED_SHUFFLE_ID ||
+              msg.shuffleMergeId == mergePartitionsInfo.shuffleMergeId;
+      AppAttemptShuffleMergeId currentAppAttemptShuffleMergeId =
+          new AppAttemptShuffleMergeId(
+              msg.appId, msg.appAttemptId, msg.shuffleId, mergePartitionsInfo.shuffleMergeId);
+      AppAttemptShuffleMergeId appAttemptShuffleMergeId = new AppAttemptShuffleMergeId(
+          msg.appId, msg.appAttemptId, msg.shuffleId, msg.shuffleMergeId);
+      if(deleteCurrent) {
+        // request to clean up shuffle we are currently hosting
+        if (!mergePartitionsInfo.isFinalized()) {
+          submitCleanupTask(() -> {
+            closeAndDeleteOutdatedPartitions(
+                currentAppAttemptShuffleMergeId, mergePartitionsInfo.shuffleMergePartitions);
+            writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
+          });
+        } else {
+          submitCleanupTask(() -> {
+            deleteMergedFiles(currentAppAttemptShuffleMergeId,
+                mergePartitionsInfo.getReduceIds());
```

Comment:
I think if mergePartitionsInfo is not finalized, `mergePartitionsInfo.getReduceIds()` will be empty. How about cleaning the current unfinalized partitions `AppShuffleMergePartitionsInfo.shuffleMergePartitions`?

Reply:
```
    if (!mergePartitionsInfo.isFinalized()) {
          submitCleanupTask(() -> {
            closeAndDeleteOutdatedPartitions(
                currentAppAttemptShuffleMergeId, mergePartitionsInfo.shuffleMergePartitions);
            writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
          });
        } else {
          submitCleanupTask(() -> {
            deleteMergedFiles(currentAppAttemptShuffleMergeId,
                mergePartitionsInfo.getReduceIds());
            writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
            mergePartitionsInfo.setReduceIds(new int[0]);
          });
        }
```

Yes, if mergePartitionsInfo is not finalized `mergePartitionsInfo.isFinalized()`, it will clean the current partitions.
It depends on the `AppShuffleMergePartitionsInfo` in `appShuffleInfo.shuffles`

Reply:
Not sure I follow ... this is in the `else` block of `!mergePartitionsInfo.isFinalized()` - so it is finalized ?
Or is this a stale comment that has already been addressed @yabola ?

Reply:
Resolving conversation, please reopen if there is something additionally missing.